### DISCUSSION TITLE
Response Body in Logger is always empty

### DISCUSSION THREAD

**Discussion by ErikPelli** [earthboundkid/requests#98] [Support] [answered]
I paste here an example code to reproduce the issue:
```go
logger := func(_ *http.Request, res *http.Response, _ error, _ time.Duration) {
		result, err := io.ReadAll(res.Body)
		log.Println("Body Length", len(result))
		log.Println("Body", result)
		log.Println("Body error", err)
	}

	// Wrap an existing transport or use nil for http.DefaultTransport
	baseTrans := http.DefaultClient.Transport
	trans := requests.LogTransport(baseTrans, logger)

	var result any
	// A random JSON API
	err := requests.URL("https://worldtimeapi.org/api/timezone").
		Transport(trans).
		ToJSON(&result).
		Fetch(context.Background())
	log.Println("Result", result)
	log.Println("Request Error", err)
```

The `res.Body` field is an empty reader but there is clearly a response body.
Actually I solved this problem by creating a custom `http.RoundTripper` and assigning it to the http client to get the raw request and response and delegating the http handling to the original  `http.DefaultTransport`, and i get the required response body data in this way.

I would suggest to add more tests in this part of code after fixing the problem.

**Comment by earthboundkid** [comment]
The response body is an io.ReadCloser, so it can't be read by both ToJSON and the logger. It needs to be one or the other. Once you read from a reader, it's gone. You can work around that by various means (save to a string and then use the string for logging and JSON or use some other kind of replayable buffer), but it's sort of a fundamental part of how io.Reader works that to save memory by default it won't "rewind".

**Reply by ErikPelli** [reply]
Wait, the transport is called before the ToJSON() function when the response is received, right?

The problem is that in the output of my code the length of the data inside resp.Body Reader (logger function argument) is 0.
But then the result is saved correctly, It's logical that if I read in the logger, then there is no result, but it shouldn't be the opposite...

**Reply by earthboundkid** [reply, marked as answer]
No, the logger transport is triggered last. That's how it knows what the request duration is.

**Reply by ErikPelli** [reply]
Mmmm ok that's unexpected for a logger.
Btw I don't care about the duration so I keep my working custom implementation that uses directly the http client so I can both log the response body and decode it later using requests.
Maybe another function that does something like this without the duration can be useful.

Thanks for the availability!
